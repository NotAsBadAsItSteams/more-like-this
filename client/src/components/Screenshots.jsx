import React from 'react';
import PropTypes from 'prop-types';

const INTERVAL_LENGTH = 1000;

class Screenshots extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      currentImageIndex: 0,
      currentImage: props.screenshots[0],
    };
    this.changeScreenshot = this.changeScreenshot.bind(this);
  }

  componentDidMount() {
    this.interval = setInterval(this.changeScreenshot, INTERVAL_LENGTH);
  }

  componentWillUnmount() {
    clearInterval(this.interval);
  }

  changeScreenshot() {
    let nextImageIndex;
    if (this.state.currentImageIndex === (this.props.screenshots.length - 1)) {
      nextImageIndex = 0;
    } else {
      nextImageIndex = this.state.currentImageIndex + 1;
    }
    this.setState({
      currentImageIndex: nextImageIndex,
      currentImage: this.props.screenshots[nextImageIndex],
    });
  }

  render() {
    return (
      <div>
        <img src={this.state.currentImage} alt="game screenshot" />
      </div>
    );
  }
}

Screenshots.propTypes = {
  screenshots: PropTypes.arrayOf(PropTypes.string).isRequired,
};

export default Screenshots;
