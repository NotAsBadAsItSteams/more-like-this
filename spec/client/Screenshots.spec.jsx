import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import Screenshots from '../../client/src/components/Screenshots';


describe('Screenshots component', () => {
  const INTERVAL_LENGTH = 1000;
  const screenshots = [
    '/assets/screenshots/12-0.png',
    '/assets/screenshots/12-1.png',
    '/assets/screenshots/12-2.png',
    '/assets/screenshots/12-3.png',
  ];
  jest.useFakeTimers();

  test('should match snapshot', () => {
    const component = renderer.create(<Screenshots screenshots={screenshots} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should change image every second', () => {
    const wrapper = mount(<Screenshots screenshots={screenshots} />);
    expect(wrapper.state('currentImageIndex')).toEqual(0);
    expect(wrapper.state('currentImage')).toEqual(screenshots[0]);
    jest.advanceTimersByTime(INTERVAL_LENGTH);
    expect(wrapper.state('currentImageIndex')).toEqual(1);
    expect(wrapper.state('currentImage')).toEqual(screenshots[1]);
    wrapper.unmount();
  });

  test('should reset back to first image when it reaches the end', () => {
    const wrapper = mount(<Screenshots screenshots={screenshots} />);
    for (let i = 0; i < screenshots.length; i += 1) {
      expect(wrapper.state('currentImageIndex')).toEqual(i);
      expect(wrapper.state('currentImage')).toEqual(screenshots[i]);
      jest.advanceTimersByTime(INTERVAL_LENGTH);
    }
    expect(wrapper.state('currentImageIndex')).toEqual(0);
    expect(wrapper.state('currentImage')).toEqual(screenshots[0]);
    wrapper.unmount();
  });

  test('should cancel the set interval when unmounted', () => {
    const wrapper = mount(<Screenshots screenshots={screenshots} />);
    expect(setInterval).toHaveBeenCalled();
    wrapper.unmount();
    expect(clearInterval).toHaveBeenCalled();
  });
});
