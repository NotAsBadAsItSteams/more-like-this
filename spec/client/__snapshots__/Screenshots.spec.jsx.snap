// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Screenshots component should match snapshot 1`] = `
<div>
  <img
    alt="game screenshot"
    src="/assets/screenshots/12-0.png"
  />
</div>
`;
